### DISCUSSION TITLE
Combine Commands/Permissions in Docs + Automate all commands/permissions into a web UI for Docs

### DISCUSSION THREAD

**Discussion by TomLewis** [Zrips/CMI#3531] [Feature Requests]
Hi!

CMI has so many features that most people will never know exist because the docs are so rarely updated, to solve this im going to request a few changes.

1) Automate All commands/permissions when you build the plugin and export all commands/permissions into a JSON file that will be uploaded somewhere (GitHub maybe) so then your Docs page on the website can pull this JSON file and format it into a table.
have an array of all your descriptions and example uses for each command that you can add to when you release something new, so your examples/descriptions are also managed via Git locally/pulled remotely. 
2) Change your Docs website to use DataTables https://datatables.net/
This can pull an external data source to populate a searchable table, you could easily combine both Commands & Permissions into one giant table that is searchable, organized and can be automated if you create something to automatically publish updated permissions/commands when you build CMI so you don't have to maintain it. 
You use bootstrap, they even have style choices specifically for bootstrap https://datatables.net/examples/styling/bootstrap4
3) Generate plugin.yml to include all permissions/commands too, so then we can extract and search locally.

For example, I just tried to setup scavenging, I cant find the perissions/use cases on the zrips website in commands or the permissions, and theres nothing in the plugin.yml or on the special page it has on the website.

A second example, I couldn't find /openbook anywhere either, I found this by mistake one day in-game, this is such an amazing feature it should be advertised to people! There are entire plugins that do this 1 thing! I have NPCs that show all my in-game tutorials through /openbook now!

**Comment by Zrips** [comment]
Well, web page is to have some what one place where you can find information when you are not ingame. What said, you have 2 commands **/cmi checkcommand (keyWord)** and **/cmi checkperm (keyWord)** which will output related commands or permissions. Which are always up to date with current plugin state, while web page could be outdated. 
it is cool idea to have intractable table, will keep that in mind

**Comment by mrfloris** [comment]
@Zrips He has a fair point that the commands/permissions are not listed on the dedicated scavenge page `https://www.zrips.net/cmi/extra/scavenge/`, the permissions page doesn't have the word scavenge on it for `cmi.command.scavenge` and `cmi.command.scavenge.others`. `https://www.zrips.net/cmi/permissions/` and the command is listed, but currently doesn't link to the dedicated scavenge page `https://www.zrips.net/cmi/commands/`.

Openbook was missing, but recently added to the commands page after my suggestion the other week.

It's true in my opinion that when you have a huge plugin and slowly building it up over the years, especially when you move documentation from spigotmc to your own site, that it's ending up as not super organized (but i do think an amazing job has been done, and it's very much complete. Just 1% is missing). 

The suggestion to streamline documenting content such as permissions and commands, is of course important. And super helpful in the long run. I think that's worth to consider yeh. 

If the commands and permissions are in plugin.yml is great, but it won't include details such as full descriptions, links to other pages, permission explanations, or what's linked to what. It can get complicated when there's around 300 commands with perhaps double the permissions and about 200 placeholders. 

Maybe a bit of a compromise can be made. 

I do like the idea of a page the community can contribute to, I always enjoyed updating the spigot wiki to make sure there's no typing errors, to complete a description with newer details, or to link to 'more info' things.  The current website doesn't permit for this at this point.

**Comment by 456dev** [comment]
Building off of what @mrfloris said, if you didn't want to use spigots wiki (i.e too crowded with other plugins, or it's not in a specific namespace or whatever) then have you thought about using githubs wiki. You can use markdown for formatting, and allow only trusted people (repository contributers) or anyone to edit the wiki

Source: [Here](https://help.github.com/en/github/building-a-strong-community/about-wikis) and [Here](https://help.github.com/en/github/building-a-strong-community/changing-access-permissions-for-wikis)

**Comment by TomLewis** [comment]
Alternativly, you can host the webpage using Github pages, then we can still use the DataTables that people can search/organize AND if we have all data for the table in JSON this could also be in the GitHub so the community could update it :)

Or you could have the JSON file just on GitHub and then that is pulled from your website into the table, I have used DataTables extensively with Bootstrap support, so if you moved your website docs to be a Github page/netlify from a Git or now.sh then the website could be community updated.
